File: dynamic programming/java/leetcode132_Palindrome_Partitioning_II.java
//Given a string s, partition s such that every substring of the partition is a palindrome.
//
//Return the minimum cuts needed for a palindrome partitioning of s.
//
//Example:
//
//Input: "aab"
//Output: 1
//Explanation: The palindrome partitioning ["aa","b"] could be produced using 1 cut.

package wenxinjie2;

public class leetcode132_Palindrome_Partitioning_II {
    public int minCut(String s) {       
        char[] c = s.toCharArray();
        int n = c.length;
        int[] cut = new int[n];
        boolean[][] pal = new boolean[n][n];
        for(int i = 0; i < n; i ++){
            int min = i ;
            for (int j = 0; j <= i; j ++){
                if ((c[i] == c[j]) && ( j + 1 > i - 1 || pal[j+1][i-1] ) ){
                    pal[j][i] = true;
                    min = j==0? 0: Math.min(min, cut[j-1] + 1);
                }
            }
            cut[i] = min;            
        }
        return cut[n-1];      
    }
}

//Time: O(n^2)
//Space: O(n^2)
//Difficulty: medium

File: dynamic programming/python/leetcode132_Palindrome_Partitioning_II.py
# Given a string s, partition s such that every substring of the partition is a palindrome.

# Return the minimum cuts needed for a palindrome partitioning of s.

# Example:

# Input: "aab"
# Output: 1
# Explanation: The palindrome partitioning ["aa","b"] could be produced using 1 cut.


class Solution(object):
    def minCut(self, s):
        """
        :type s: str
        :rtype: int
        """
        n = len(s)
        dp = [0 for _ in range(n)]
        pal = [[False for i in range(n)] for j in range(n)]
        for i in range(n):
            cut = i
            for j in range(i+1):
                if s[i] == s[j] and (j + 1 > i - 1 or pal[j+1][i-1]) :
                    pal[j][i] = True
                    cut = 0 if j == 0 else min(dp[j] + 1, cut)
            dp[i] = cut
        return dp[-1]

# Time: O(n^2)
# Space: O(n^2)
# Difficulty: medium